src/wallet/wallet.cpp

```diff
@@ -1902,95 +1909,123 @@ void CWallet::ResendWalletTransactions(int64_t nBestBlockTime, CConnman* connman
  * @{
  */
 
+void CWallet::MarkBalancesDirty()
+{
+    fBalancesDirty = true;
+    BalancesDidChange();
+}
 
 CAmount CWallet::GetBalance() const
 {
-    CAmount nTotal = 0;
+    if (!fBalancesDirty) {
+        return nBalanceCache;
```

Comment:
I think think you need to acquire cs_wallet lock here, otherwise this could return a partial sum if `GetBalance` got called from two threads at the same time and one thread returned `nBalanceCache` while the other thread was in the middle of the `nBalanceCache +=` loop.

Alternately, you could bring back the nTotal local variable and just set `nBalanceCache = nTotal`  atomically.